graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,376 @@
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldParameters;
+import graphql.schema.*;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class CacheControlInstrumentation extends SimpleInstrumentation {
+    private final int defaultMaxAge;
+    private final HintCache hintCache = new HintCache();
+
+    static private final String DIRECTIVE_NAME = "cacheControl";
+    static private final String EXTENSION_KEY = "cache-control";
+
+    public CacheControlInstrumentation() {
+        this(0);
+    }
+
+    public CacheControlInstrumentation(int defaultMaxAge) {
+        this.defaultMaxAge = defaultMaxAge;
+    }
+
+    @Override
+    public InstrumentationState createState() {
+        return new CacheControlState();
+    }
+
+    @Override
+    public InstrumentationContext<ExecutionResult> beginField(InstrumentationFieldParameters parameters) {
+        CacheControlState state = parameters.getInstrumentationState();
+        CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+        boolean inheritMaxAge = false;
+
+        GraphQLOutputType returnType = parameters.getExecutionStepInfo().getUnwrappedNonNullType();
+
+        // There's no way to set a cacheControl directive on the _entities field or the _Entity union in SDL.
+        // Instead, we can determine the possible return types from the representations arguments and select the
+        // most restrictive cache policy from that set.
+
+        if (isEntityField(returnType)) {
+            Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+            if (representations instanceof List) {
+                typesFromEntitiesArgument(representations, parameters.getExecutionContext().getGraphQLSchema())
+                        .stream()
+                        .map(type -> hintCache.getHint((GraphQLCompositeType) type))
+                        .filter(Optional::isPresent)
+                        .forEach(hint -> fieldPolicy.restrict(hint.get()));
+            }
+        }
+
+        if (returnType instanceof GraphQLList) {
+            GraphQLType wrappedType = ((GraphQLList) returnType).getWrappedType();
+            if (wrappedType instanceof GraphQLUnionType) {
+                if (((GraphQLUnionType) wrappedType).getName().equals(_Entity.typeName)) {
+                    System.out.printf("found type: %s%n", wrappedType);
+                }
+            }
+        }
+
+        // Cache hint on the return type of this field if it's a composite type
+
+        if (returnType instanceof GraphQLList) {
+            returnType = (GraphQLOutputType) ((GraphQLList) returnType).getWrappedType();
```

Comment:
Should that use `GraphQLTypeUtil.unwrapAll(returnType);` instead? In case the return type is a list of list of nullable, etc... ?

Reply:
Yesss thank you I could not find that in the docs!